### PR TITLE
feat(sprite)!: make coverageGamma & per-sprite uvOffset opt-in (import-triggered)

### PR DESCRIPTION
**Pull request by ryantrem**
## Summary

Two optional `Sprite2DLayer` features — **coverage gamma** (glyph "stem darkening") and **per-sprite UV offset** (uvScroll parallax) — were paying their cost on *every* sprite scene, even ones that never used them. Their plumbing (shader permutations, an extra vertex attribute + widened instance stride, a per-frame UBO write, pipeline-cache key parts, and the public option/field strings) lived in the always-loaded sprite path.

This PR converts both to the **import-trigger pattern** already used across the engine (`setFog`, `setClipPlane`, `createSprite2DCustomShader`, `enableBoneControl`, …): the feature code is pulled into the bundle **only when you import the function that uses it**. Sprite scenes that never touch these features ship none of the code.

## Why

Per the repo's core pillars (100% tree-shakable, minimal footprint, "always use extensions / never hardcode"), an optional feature should cost **zero bytes when unused**. These two were the remaining value-gated sprite features still hardcoded into the always-loaded path. Routing them through the same opt-in mechanism keeps the always-loaded sprite pipeline lean and matches how every other optional feature in the engine already works.

## What changed

**Coverage gamma** — was a create-time option (`Sprite2DLayerOptions.coverageGamma`) + an always-loaded public `coverageGamma` field.
- New public setter **`setSprite2DCoverageGamma(layer, gamma)`** is now the only way to enable it.
- The value is stored on an `@internal` field (set *only* via the setter), so callers can't write a value the renderer would silently ignore.
- The `pow(...)` fragment permutation, the `aa.x` UBO write, and the pipeline-key part now live behind a null-by-default hook (`sprite-coverage-gamma-hook.ts`) in a tree-shaken module (`sprite-2d-coverage-gamma.ts`) — mirroring the custom-shader hook.

**Per-sprite UV offset (uvScroll)** — was a create-time option (`Sprite2DLayerOptions.uvScroll`) + an add-time prop (`Sprite2DProps.uvOffset`).
- Now lazily enabled on the **first `setSprite2DUvOffset(layer, index, uvOffset)` call**: it widens the layer's instance layout (an O(count) one-time re-stride of existing sprites) and flips it to the wide layout.
- The narrow→wide re-stride lives in a tree-shaken module (`sprite-2d-uvscroll.ts`). The always-loaded pipeline needs **no code** from it: the module precomputes the `@location(7)` vertex attribute and stashes it on the layer as plain data, so the pipeline just *consumes* the data (no hook indirection needed).

**Latent correctness fix** — `ensureSpriteInstanceBuffer` now compares **byte size** (`currentBuffer.size`) instead of sprite count, so a lazy stride-widen correctly triggers a GPU instance-buffer reallocation + full re-upload. Without this, enabling uvScroll after sprites were added could upload wide data into a too-small buffer.

## Bundle-size impact

This is the headline trade, and it is intentional:

- **Scenes that do NOT use these features are slightly *smaller* than before** — the gamma + uvScroll plumbing no longer ships to them.
- **Scenes that DO use per-sprite UV offset are slightly *larger* than before** — the feature code now lives in their bundle (it just no longer leaks into everyone else's).

| Scene | Uses uvOffset? | Before (KB raw) | After (KB raw) | Δ |
|------|----|------|------|------|
| scene50 | no | 16.1 | 15.7 | **−0.4** |
| scene51 | no | 16.6 | 16.1 | **−0.5** |
| scene52 | no | 57.8 | 57.4 | **−0.4** |
| scene53 | no | 58.0 | 57.5 | **−0.5** |
| scene58 | no | 19.3 | 18.8 | **−0.5** |
| scene92 | no | 19.1 | 18.6 | **−0.5** |
| scene93 | no | 20.2 | 19.8 | **−0.4** |
| scene97 | no | 16.3 | 15.8 | **−0.5** |
| **scene96** | **yes** | **15.8** | **15.9** | **+0.1** |

(The per-scene `lab/public/bundle/manifest/` files are regenerated and committed.)

## Breaking changes

The create-time options and add-time prop were removed in favor of the setters.

```ts
// coverage gamma
- const layer = createSprite2DLayer(atlas, { coverageGamma: 2 });
+ const layer = createSprite2DLayer(atlas);
+ setSprite2DCoverageGamma(layer, 2);

// per-sprite UV offset (uvScroll parallax)
- const layer = createSprite2DLayer(atlas, { uvScroll: true });
- addSprite2DIndex(layer, { positionPx, sizePx, frame, uvOffset: [u, v] });
+ const layer = createSprite2DLayer(atlas);
+ const i = addSprite2DIndex(layer, { positionPx, sizePx, frame });
+ setSprite2DUvOffset(layer, i, [u, v]); // first call enables scroll
```

`setSprite2DUvOffset` already existed (it was the live-update setter); it is now also the *enable* trigger, so most call sites change little. `Sprite2DLayerOptions.coverageGamma`, `Sprite2DLayerOptions.uvScroll`, and `Sprite2DProps.uvOffset` are gone from the public API.

BREAKING CHANGE: `Sprite2DLayerOptions.coverageGamma`, `Sprite2DLayerOptions.uvScroll`, and `Sprite2DProps.uvOffset` have been removed. Enable coverage gamma via `setSprite2DCoverageGamma(layer, gamma)` and per-sprite UV offset via `setSprite2DUvOffset(layer, index, uvOffset)` (the first call enables uvScroll and widens the layer).

## Validation

- **Parity:** all sprite parity scenes pass with unchanged MAD; scene96 (uvScroll) MAD = 0.000.
- **Unit tests:** sprite unit suites pass; the uvScroll unit block was rewritten for the lazy-enable contract (narrow until first `setSprite2DUvOffset`, then re-strides existing sprites).
- **Build:** full package build + d.ts trim clean — public API exposes only the setters; internal fields are trimmed.
- **Manual:** exercised the platformer demo (the real per-frame `setSprite2DUvOffset` parallax path) — multi-band parallax scrolls cleanly with no rendering artifacts.

### CODE REVIEW

**Comment by bjsplat** [comment]
## API Changes

API Extractor detected public API changes for `@babylonjs/lite`.

**Potentially breaking changes detected.** Removed or changed public API lines:

- `coverageGamma: number;`
- `coverageGamma?: number;`
- `uvScroll?: boolean;`
- `uvOffset?: [number, number];`

<details>
<summary>API Extractor diff</summary>

```diff
diff --git a/home/vsts/work/1/s/test-results/api-report/target/temp/babylon-lite.api.md b/home/vsts/work/1/s/test-results/api-report/current/temp/babylon-lite.api.md
index c6ff1fff..a1579f8c 100644
--- a/home/vsts/work/1/s/test-results/api-report/target/temp/babylon-lite.api.md
+++ b/home/vsts/work/1/s/test-results/api-report/current/temp/babylon-lite.api.md
@@ -4649,6 +4649,9 @@ export function setSpatialOrientation(host: AudioGraphHost, orientation: Vec3):
 // @public
 export function setSpatialPosition(host: AudioGraphHost, position: Vec3): void;
 
+// @public
+export function setSprite2DCoverageGamma(layer: Sprite2DLayer, gamma: number): void;
+
 // @public
 export function setSprite2DFrame(handle: Sprite2DHandle, frame: number): void;
 
@@ -5065,7 +5068,6 @@ export interface Sprite2DLayer {
     readonly blendMode: SpriteBlendMode;
     // (undocumented)
     readonly count: number;
-    coverageGamma: number;
     readonly customShader?: Sprite2DCustomShader;
     // (undocumented)
     readonly depth: Sprite2DDepthMode;
@@ -5088,7 +5090,6 @@ export interface Sprite2DLayerOptions {
     blendMode?: SpriteBlendMode;
     // (undocumented)
     capacity?: number;
-    coverageGamma?: number;
     customShader?: Sprite2DCustomShader;
     // (undocumented)
     depth?: Sprite2DDepthMode;
@@ -5098,7 +5099,6 @@ export interface Sprite2DLayerOptions {
     // (undocumented)
     order?: number;
     pivot?: [number, number];
-    uvScroll?: boolean;
     // (undocumented)
     view?: Partial<Sprite2DView>;
     // (undocumented)
@@ -5123,7 +5123,6 @@ export interface Sprite2DProps {
     rotation?: number;
     // (undocumented)
     sizePx?: [number, number];
-    uvOffset?: [number, number];
     // (undocumented)
     visible?: boolean;
     z?: number;
```

</details>

**Comment by bjsplat** [comment]
## Bundle Size Changes

### Decreases

| Package | Current | Master | Change |
|---------|---------|--------|--------|
| Scene 51 — Soft-Edged Sprite Grid (Premultiplied)<br/>`scene51` | 16 KB | 17 KB | -1 KB |
| Scene 52 — HUD on 3D<br/>`scene52` | 57 KB | 58 KB | -1 KB |

*Sizes rounded to nearest KB. Run `pnpm build:bundle-scenes` locally to verify.*

**Comment by bjsplat** [comment]
### Lab - Static Site

[Open deployed site](https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/lite/20260626.5/lab/index.html)

<sub>Build 20260626.5 - merge @ 5b227cd5a7eeb58ccdd8a08df3bdb599ae6b9980</sub>

File: packages/babylon-lite/src/sprite/sprite-2d-coverage-gamma.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        perDevice.set(key, module);
+    }
+    return module;
+}
+
+/** True when `layer` has an active (non-identity) coverage gamma set via the opt-in setter. */
+function isGammaActive(layer: Sprite2DLayer): boolean {
+    const g = layer._coverageGamma;
+    return g != null && g !== 1;
+}
```
[the remark this answers is not in the document]
Comment (ryantrem) [reply packages/babylon-lite/src/sprite/sprite-2d-coverage-gamma.ts]: Good catch. Fixed in be0a533: `isGammaActive` now requires a **finite, positive, non-identity** value — `g != null && Number.isFinite(g) && g > 0 && g !== 1`. That rejects `0`, negatives, `NaN`, and `Infinity`, so `writeUbo` can never feed a non-finite `1/gamma` exponent into `pow(s.a, L.aa.x)`; those values now fall back to the base (non-gamma) shader with `aa.x = 0`. The setter docstring documents this, and the new edge-case unit test asserts `ubo[12] === 0` and base-shader selection for each of `1, 0, -2, NaN, Infinity`.

File: packages/babylon-lite/src/sprite/sprite-2d-coverage-gamma.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * calling it later is also safe — the renderer re-fetches the pipeline each frame and rebuilds it
+ * with the gamma permutation on the next frame.
+ *
+ * @param layer - The sprite layer to configure.
+ * @param gamma - Coverage gamma. Typical text values are ~1.8–2.2; `1` disables the effect.
+ */
+export function setSprite2DCoverageGamma(layer: Sprite2DLayer, gamma: number): void {
+    _registerSpriteCoverageGammaHook(COVERAGE_GAMMA_HOOK);
+    (layer as { _coverageGamma?: number })._coverageGamma = gamma;
+}
```
[the remark this answers is not in the document]
Comment (ryantrem) [reply packages/babylon-lite/src/sprite/sprite-2d-coverage-gamma.ts:109]: Added in be0a533. New `coverageGamma (opt-in via setSprite2DCoverageGamma)` describe block in `sprite-renderer.test.ts` covering the full contract:
- setter stores the internal `_coverageGamma` value; plain layers never name it;
- a gamma layer builds a **distinct pipeline + extra shader module** (the `cg` key part differs) and the gamma fragment includes the `pow(s.a, L.aa.x)` permutation while the base fragment does not; re-requesting hits the cache;
- `writeUbo` puts `aa.x = 1/gamma` in UBO slot `[12]` for an active layer (e.g. gamma 2 → 0.5);
- the edge-case / identity values write `0` to slot `[12]` and select the base shader;
- a plain layer leaves slot `[12]` at 0 once the hook is registered.

Suite is green (52 tests).

**Comment by bjsplat** [comment]
## API Changes

API Extractor detected public API changes for `@babylonjs/lite`.

**Potentially breaking changes detected.** Removed or changed public API lines:

- `coverageGamma: number;`
- `coverageGamma?: number;`
- `uvScroll?: boolean;`
- `uvOffset?: [number, number];`

<details>
<summary>API Extractor diff</summary>

```diff
diff --git a/home/vsts/work/1/s/test-results/api-report/target/temp/babylon-lite.api.md b/home/vsts/work/1/s/test-results/api-report/current/temp/babylon-lite.api.md
index c6ff1fff..a1579f8c 100644
--- a/home/vsts/work/1/s/test-results/api-report/target/temp/babylon-lite.api.md
+++ b/home/vsts/work/1/s/test-results/api-report/current/temp/babylon-lite.api.md
@@ -4649,6 +4649,9 @@ export function setSpatialOrientation(host: AudioGraphHost, orientation: Vec3):
 // @public
 export function setSpatialPosition(host: AudioGraphHost, position: Vec3): void;
 
+// @public
+export function setSprite2DCoverageGamma(layer: Sprite2DLayer, gamma: number): void;
+
 // @public
 export function setSprite2DFrame(handle: Sprite2DHandle, frame: number): void;
 
@@ -5065,7 +5068,6 @@ export interface Sprite2DLayer {
     readonly blendMode: SpriteBlendMode;
     // (undocumented)
     readonly count: number;
-    coverageGamma: number;
     readonly customShader?: Sprite2DCustomShader;
     // (undocumented)
     readonly depth: Sprite2DDepthMode;
@@ -5088,7 +5090,6 @@ export interface Sprite2DLayerOptions {
     blendMode?: SpriteBlendMode;
     // (undocumented)
     capacity?: number;
-    coverageGamma?: number;
     customShader?: Sprite2DCustomShader;
     // (undocumented)
     depth?: Sprite2DDepthMode;
@@ -5098,7 +5099,6 @@ export interface Sprite2DLayerOptions {
     // (undocumented)
     order?: number;
     pivot?: [number, number];
-    uvScroll?: boolean;
     // (undocumented)
     view?: Partial<Sprite2DView>;
     // (undocumented)
@@ -5123,7 +5123,6 @@ export interface Sprite2DProps {
     rotation?: number;
     // (undocumented)
     sizePx?: [number, number];
-    uvOffset?: [number, number];
     // (undocumented)
     visible?: boolean;
     z?: number;
```

</details>

**Comment by bjsplat** [comment]
## Bundle Size Changes

### Decreases

| Package | Current | Master | Change |
|---------|---------|--------|--------|
| Scene 51 — Soft-Edged Sprite Grid (Premultiplied)<br/>`scene51` | 16 KB | 17 KB | -1 KB |
| Scene 52 — HUD on 3D<br/>`scene52` | 57 KB | 58 KB | -1 KB |

*Sizes rounded to nearest KB. Run `pnpm build:bundle-scenes` locally to verify.*

**Comment by bjsplat** [comment]
### Lab - Static Site

[Open deployed site](https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/lite/20260626.6/lab/index.html)

<sub>Build 20260626.6 - merge @ 3efd31720a554b599235d3346cf98607bff2ff6f</sub>

**Review by deltakosh — APPROVED** [review]
